hermes-common/src/main/java/pl/allegro/tech/hermes/common/message/wrapper/AvroMessageSchemaVersionTruncationContentWrapper.java

```diff
@@ -0,0 +1,89 @@
+    public AvroMessageSchemaVersionTruncationContentWrapper(SchemaRepository schemaRepository,
+                                                            AvroMessageContentWrapper avroMessageContentWrapper,
+                                                            DeserializationMetrics deserializationMetrics,
+                                                            ConfigFactory configFactory) {
+        this(
+            schemaRepository,
+            avroMessageContentWrapper,
+            deserializationMetrics,
+            configFactory.getBooleanProperty(Configs.SCHEMA_VERSION_TRUNCATION_ENABLED)
+        );
+    }
+
+    public AvroMessageSchemaVersionTruncationContentWrapper(SchemaRepository schemaRepository,
+                                                            AvroMessageContentWrapper avroMessageContentWrapper,
+                                                            DeserializationMetrics deserializationMetrics,
+                                                            boolean magicByteTruncationEnabled) {
+        this.schemaRepository = schemaRepository;
+        this.avroMessageContentWrapper = avroMessageContentWrapper;
+        this.magicByteTruncationEnabled = magicByteTruncationEnabled;
+
+        this.deserializationErrorsWithSchemaVersionTruncation = deserializationMetrics.errorsForSchemaVersionTruncation();
+        this.deserializationWithSchemaVersionTruncation = deserializationMetrics.usingSchemaVersionTruncation();
+    }
+
+
+    @Override
+    public AvroMessageContentUnwrapperResult unwrap(byte[] data, Topic topic, Integer schemaId, Integer schemaVersion) {
+        try {
+            deserializationWithSchemaVersionTruncation.inc();
+
+            byte[] dataWithoutMagicByteAndSchemaId = trimMagicByteAndSchemaVersion(data);
+            CompiledSchema<Schema> avroSchema = schemaRepository.getAvroSchema(topic, SchemaVersion.valueOf(schemaVersion));
+
+            return AvroMessageContentUnwrapperResult.success(avroMessageContentWrapper.unwrapContent(dataWithoutMagicByteAndSchemaId, avroSchema));
+        } catch (Exception e) {
+            logger.warn(
+                    "Could not unwrap content for topic [{}] using schema id provided in header [{}] - falling back",
+                    topic.getQualifiedName(), schemaVersion, e);
+
+            deserializationErrorsWithSchemaVersionTruncation.inc();
+
+            return AvroMessageContentUnwrapperResult.failure();
+        }
+    }
+
+    @Override
+    public boolean isApplicable(byte[] data, Topic topic, Integer schemaId, Integer schemaVersion) {
+        return magicByteTruncationEnabled && containsSchemaVersionInMagicByteAndInHeader(data, schemaVersion);
+    }
+
+    private boolean containsSchemaVersionInMagicByteAndInHeader(byte[] data, Integer schemaVersion) {
+        return SchemaAwareSerDe.startsWithMagicByte(data) && schemaVersion != null;
+    }
+
+    private byte[] trimMagicByteAndSchemaVersion(byte[] data) {
```

Comment:
What do you think about moving this method to the `SchemaAwareSerDe`? Thanks to this we will keep the format of the magic byte header encapsulated inside that class.

Reply:
Good idea, I will move it there.